### PR TITLE
[14.0][IMP] sale_order_line_chained_move: do not merge propagated moves

### PR DESCRIPTION
**Pull request by HekkiMelody**
This small change means that even in stock moves that are propagated from the initial one, moves for the same product are not merged if they refer to different sale order lines.

Without this change, the moves are merged and the field `related_sale_line_id` is overwritten with the last move processed, losing the link to the other sale order lines.

### CODE REVIEW

**Comment by HekkiMelody** [comment]
I am isolating the module with incompatible tests `sale_stock_line_customer_ref` into its own set of tests. 

(I could have isolated this module as well, but this module has dependencies, whereas `sale_stock_line_customer_ref` does not)

**Comment by francesco-ooops** [comment]
@rousseldenis what do you think?

**Comment by elvise** [comment]
I think we need to IMP with an option like this:
IMG_0959

**Comment by HekkiMelody** [comment]
> I think we need to IMP with an option like this: IMG_0959

Implemented this. The default for all rules is that they are kept separate.

**Comment by francesco-ooops** [comment]
@rousseldenis ok, I think we've made good progress. 

Would this solution work for you? Do you have suggestions? thanks :)

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

**Review by ilyasProgrammer — APPROVED** [review]
LGTM

**Review by geomer198 — APPROVED** [review]
LGTM!
Could you add message for each assertion in tests?

**Comment by HekkiMelody** [comment]
done!

**Comment by francesco-ooops** [comment]
@rousseldenis do you think it's good for merge? thanks!

**Comment by francesco-ooops** [comment]
@rousseldenis gentle reminder :)

**Comment by elvise** [comment]
@rousseldenis gentle reminder ;)

**Comment by francesco-ooops** [comment]
@OCA/crm-sales-marketing-maintainers would it be possible to merge? approved since 3 weeks. Thanks!

**Comment by francesco-ooops** [comment]
@pedrobaeza can we go?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -50,11 +50,18 @@ jobs:
             include: "sale_procurement_group_by_commitment_date"
             name: test with OCB
           - container: ghcr.io/oca/oca-ci/py3.6-odoo14.0:latest
-            exclude: "sale_order_line_sequence,sale_procurement_group_by_commitment_date"
+            include: "sale_stock_line_customer_ref"
```
Comment (pedrobaeza) [inline .github/workflows/test.yml]: You mustn't touch this file manually, but through copier.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Anyway, adding yet another builds to separate modules doesn't seem to me as a good practice (other think on contrary). Instead, the code should inhibit its conflicting behavior if testing other modules. That can be done through a context and checking `config["test_enable"]`.

File: .github/workflows/test.yml (hunk 1 above)
Comment (HekkiMelody) [reply .github/workflows/test.yml]: I didn't know, thanks for pointing it out.

**Comment by simahawk** [comment]
> Anyway, adding yet another builds to separate modules doesn't seem to me as a good practice (other think on contrary). Instead, the code should inhibit its conflicting behavior if testing other modules. That can be done through a context and checking `config["test_enable"]`.

A better option, from my POV, is to keep new features OFF by default, provide a global conf for enabling them  and enable them for testing only.

This way you gain 2 things:
1. the feature can be turned on/off fast w/o uninstalling the module (which might cause other issues)
2. no need for test specific code

File: .github/workflows/test.yml (hunk 1 above)
Comment (simahawk) [reply .github/workflows/test.yml]: to be rolled back if possible, see comments above

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by francesco-ooops** [comment]
> > Anyway, adding yet another builds to separate modules doesn't seem to me as a good practice (other think on contrary). Instead, the code should inhibit its conflicting behavior if testing other modules. That can be done through a context and checking `config["test_enable"]`.
> 
> A better option, from my POV, is to keep new features OFF by default, provide a global conf for enabling them and enable them for testing only.
> 
> This way you gain 2 things:
> 
> 1. the feature can be turned on/off fast w/o uninstalling the module (which might cause other issues)
> 2. no need for test specific code

I also like this approach from a functional POV in order to maintain established flow on databases where module is already installed

@aleuffre we should probably change this field to "Merge propagated moves"

image

**Comment by HekkiMelody** [comment]
> > Anyway, adding yet another builds to separate modules doesn't seem to me as a good practice (other think on contrary). Instead, the code should inhibit its conflicting behavior if testing other modules. That can be done through a context and checking `config["test_enable"]`.
> 
> A better option, from my POV, is to keep new features OFF by default, provide a global conf for enabling them and enable them for testing only.
> 
> This way you gain 2 things:
> 
>     1. the feature can be turned on/off fast w/o uninstalling the module (which might cause other issues)
> 
>     2. no need for test specific code

I don't mind implementing this solution, but the reason why I initially activated the option by default is that, in my opinion, without it the module works in a way that is really misleading. As moves with the same product are merged, the field `related_sale_line_id` only points to one of the possible sale order lines the move originated from, while losing the link to the others.

**Comment by francesco-ooops** [comment]
@rousseldenis what do you think?

@pedrobaeza @simahawk I agree this is an IMP bordering on a FIX

**Comment by pedrobaeza** [comment]
But why do you need to separate CIs with this IMP/FIX?

**Comment by HekkiMelody** [comment]
I've reverted the changes to the CI, and turned the option of not merging moves off by default.

**Comment by HekkiMelody** [comment]
I don't think the pre-commit failure is due to my code. Possibly due to the update to python 3.12 in the CI.

**Comment by francesco-ooops** [comment]
@aleuffre rebase pls

**Comment by francesco-ooops** [comment]
@simahawk @pedrobaeza ping :)

**Review by pedrobaeza — APPROVED** [review]

**Review by simahawk — APPROVED** [review]

**Pull request merged**